File: protocol/private-balance/crates/protocol/src/action.rs
use crate::constants::{DOMAIN_ACTION, DOMAIN_ASSET};
use crate::encoding::{
    encode_address, encode_domain, encode_optional_address, encode_u16_be, encode_u64_be,
};
use crate::encryption::OutputPackage;
use crate::field::field_id;
use crate::field::is_canonical_field;
use alloc::vec::Vec;

pub fn compute_asset_field(asset: (u8, [u8; 32])) -> [u8; 32] {
    let mut bytes = Vec::with_capacity(33);
    encode_address(asset.0, &asset.1, &mut bytes).unwrap();
    field_id(DOMAIN_ASSET, &bytes)
}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum ActionKind {
    Deposit = 1,
    PrivateTransfer = 2,
    Withdraw = 3,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Action {
    pub protocol_version: u16,
    pub kind: ActionKind,
    pub asset: (u8, [u8; 32]),
    pub action_nonce: [u8; 32],
    pub anchor_root: [u8; 32],
    pub nullifiers: [[u8; 32]; 2],
    pub outputs: [OutputPackage; 2],
    pub public_value: u64,
    pub deposit_source: Option<(u8, [u8; 32])>,
    pub public_recipient: Option<(u8, [u8; 32])>,
    pub relayer_fee: u64,
    pub relayer: Option<(u8, [u8; 32])>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ActionError {
    InvalidProtocolVersion,
    InvalidKind,
    InvalidPublicValue,
    InvalidDepositSource,
    InvalidPublicRecipient,
    InvalidSlots,
    DuplicateNullifiers,
    DuplicateOutputs,
    NonCanonicalField,
}

impl Action {
    pub fn validate_public_shape(&self) -> Result<(), ActionError> {
        const ZERO: [u8; 32] = [0; 32];
        if self.protocol_version != crate::constants::PROTOCOL_VERSION {
            return Err(ActionError::InvalidProtocolVersion);
        }
        if self.asset.0 != 1 || self.asset.1 == ZERO {
            return Err(ActionError::InvalidKind);
        }
        if self.public_value > i64::MAX as u64 || self.relayer_fee > i64::MAX as u64 {
            return Err(ActionError::InvalidPublicValue);
        }
        if !is_canonical_field(&self.anchor_root)
            || self
                .nullifiers
                .iter()
                .any(|field| !is_canonical_field(field))
            || self
                .outputs
                .iter()
                .any(|output| !is_canonical_field(&output.cm))
        {
            return Err(ActionError::NonCanonicalField);
        }
        if self.nullifiers.iter().any(|field| *field == ZERO)
            || self.outputs.iter().any(|output| {
                output.cm == ZERO
                    || output.recipient_envelope == [0; 181]
                    || output.outgoing_envelope == [0; 157]
            })
        {
            return Err(ActionError::InvalidSlots);
        }
        if self.nullifiers[0] == self.nullifiers[1] {
            return Err(ActionError::DuplicateNullifiers);
        }
        if self.outputs[0].cm == self.outputs[1].cm {
            return Err(ActionError::DuplicateOutputs);
        }
        match self.kind {
            ActionKind::Deposit => {
                if self.anchor_root != ZERO
                    || self.public_value == 0
                    || self.deposit_source.is_none()
                    || self.public_recipient.is_some()
                    || self.relayer_fee != 0
                    || self.relayer.is_some()
                {
                    return Err(ActionError::InvalidDepositSource);
                }
            }
            ActionKind::PrivateTransfer => {
                if self.anchor_root == ZERO
                    || self.public_value != 0
                    || self.deposit_source.is_some()
                    || self.public_recipient.is_some()
                    || self.relayer.is_none()
                {
                    return Err(ActionError::InvalidSlots);
                }
            }
            ActionKind::Withdraw => {
                if self.anchor_root == ZERO
                    || self.public_value == 0
                    || self.deposit_source.is_some()
                    || self.public_recipient.is_none()
                    || self.relayer.is_none()
                {
                    return Err(ActionError::InvalidPublicRecipient);
                }
            }
        }
        Ok(())
    }

    pub fn serialize_canonical_action_bytes(
        &self,
        network_id: &[u8; 32],
        realm_id: &[u8; 32],
        pool_id: &[u8; 32],
    ) -> Vec<u8> {
        let mut buf = Vec::new();
        encode_domain(DOMAIN_ACTION, &mut buf);
        encode_u16_be(self.protocol_version, &mut buf);
        buf.extend_from_slice(network_id);
        buf.extend_from_slice(realm_id);
        buf.extend_from_slice(pool_id);
        buf.push(self.kind as u8);
        encode_address(self.asset.0, &self.asset.1, &mut buf).unwrap();
        buf.extend_from_slice(&self.action_nonce);
        buf.extend_from_slice(&self.anchor_root);
        buf.extend_from_slice(&self.nullifiers[0]);
        buf.extend_from_slice(&self.nullifiers[1]);
        buf.extend_from_slice(&self.outputs[0].serialize());
        buf.extend_from_slice(&self.outputs[1].serialize());
        encode_u64_be(self.public_value, &mut buf);
        encode_u64_be(self.relayer_fee, &mut buf);
        encode_optional_address(self.relayer, &mut buf).unwrap();
        encode_optional_address(self.deposit_source, &mut buf).unwrap();
        encode_optional_address(self.public_recipient, &mut buf).unwrap();
        buf
    }

    pub fn compute_action_field(
        &self,
        network_id: &[u8; 32],
        realm_id: &[u8; 32],
        pool_id: &[u8; 32],
    ) -> [u8; 32] {
        let bytes = self.serialize_canonical_action_bytes(network_id, realm_id, pool_id);
        field_id(DOMAIN_ACTION, &bytes)
    }

    pub fn compute_asset_field(&self) -> [u8; 32] {
        compute_asset_field(self.asset)
    }

    pub fn compute_public_signals(
        &self,
        context_field: &[u8; 32],
        network_id: &[u8; 32],
        realm_id: &[u8; 32],
        pool_id: &[u8; 32],
    ) -> [[u8; 32]; 11] {
        let asset_field = self.compute_asset_field();
        self.compute_public_signals_with_asset_field(
            context_field,
            network_id,
            realm_id,
            pool_id,
            &asset_field,
        )
    }

    pub fn compute_public_signals_with_asset_field(
        &self,
        context_field: &[u8; 32],
        network_id: &[u8; 32],
        realm_id: &[u8; 32],
        pool_id: &[u8; 32],
        asset_field: &[u8; 32],
    ) -> [[u8; 32]; 11] {
        let action_field = self.compute_action_field(network_id, realm_id, pool_id);

        let mut kind_field = [0u8; 32];
        kind_field[31] = self.kind as u8;

        let mut val_field = [0u8; 32];
        val_field[24..32].copy_from_slice(&self.public_value.to_be_bytes());
        let mut relayer_fee_field = [0u8; 32];
        relayer_fee_field[24..32].copy_from_slice(&self.relayer_fee.to_be_bytes());

        [
            *context_field,
            *asset_field,
            kind_field,
            self.anchor_root,
            val_field,
            relayer_fee_field,
            action_field,
            self.nullifiers[0],
            self.nullifiers[1],
            self.outputs[0].cm,
            self.outputs[1].cm,
        ]
    }
}

File: protocol/private-balance/packages/browser/src/action.ts
import { encodeDomain, encodeU16Be, encodeU32Be, encodeU64Be } from './encoding.js';
import { fieldId, bigintTo32Bytes, isCanonicalField } from './field.js';
import { equalBytes } from './hash.js';

export const DOMAIN_ACTION = 'SKSB_ACTION_V1';
export const DOMAIN_ASSET = 'SKSB_ASSET_V1';
const MAX_PUBLIC_VALUE = (1n << 63n) - 1n;

export enum ActionKind {
  Deposit = 1,
  PrivateTransfer = 2,
  Withdraw = 3,
  FullInputExit = 4,
}

export interface OutputPackageModel {
  cm: Uint8Array; // 32 bytes
  recipientEnvelope: Uint8Array; // 181 bytes
  outgoingEnvelope: Uint8Array; // 157 bytes
}

export interface ActionModel {
  protocolVersion: number;
  kind: ActionKind;
  assetIndex?: number;
  asset?: { kind: number; payload: Uint8Array };
  actionNonce: Uint8Array;
  anchorRoot: Uint8Array;
  nullifiers: [Uint8Array, Uint8Array];
  outputs: [OutputPackageModel, OutputPackageModel, OutputPackageModel];
  publicValue: bigint;
  depositSource?: { kind: number; payload: Uint8Array };
  publicRecipient?: { kind: number; payload: Uint8Array };
}

function isZero(bytes: Uint8Array): boolean {
  return bytes.every((byte) => byte === 0);
}

function requireLength(name: string, bytes: Uint8Array, length: number): void {
  if (bytes.length !== length) throw new Error(`${name} must be ${length} bytes`);
}

function validateAddress(addr: { kind: number; payload: Uint8Array }, name: string): void {
  if (addr.kind !== 0 && addr.kind !== 1) throw new Error(`${name} kind is invalid`);
  requireLength(`${name} payload`, addr.payload, 32);
  if (isZero(addr.payload)) throw new Error(`${name} payload must be nonzero`);
}

function encodeOptionalAddress(addr: { kind: number; payload: Uint8Array } | undefined, out: number[]): void {
  if (!addr) {
    out.push(0); // absent
    for (let i = 0; i < 33; i++) out.push(0);
  } else {
    validateAddress(addr, 'Address');
    out.push(1); // present
    out.push(addr.kind);
    for (const b of addr.payload) out.push(b);
  }
}

function validateAssetBoundary(action: ActionModel): void {
  const hasAsset = action.asset !== undefined;
  const hasIndex = action.assetIndex !== undefined;
  if (hasAsset !== hasIndex) throw new Error('Boundary asset and asset index must appear together');
  if (!hasAsset) return;
  validateAddress(action.asset!, 'Asset');
  if (action.asset!.kind !== 1) throw new Error('Asset must be a contract address');
  if (!Number.isInteger(action.assetIndex) || action.assetIndex! < 0 || action.assetIndex! > 0xffff_ffff) {
    throw new Error('Asset index must be an unsigned 32-bit integer');
  }
}

function validateAction(action: ActionModel): void {
  if (action.protocolVersion !== 2) throw new Error('Unsupported action protocol version');
  if (![ActionKind.Deposit, ActionKind.PrivateTransfer, ActionKind.Withdraw, ActionKind.FullInputExit].includes(action.kind)) {
    throw new Error('Invalid action kind');
  }
  validateAssetBoundary(action);
  requireLength('Action nonce', action.actionNonce, 32);
  requireLength('Anchor root', action.anchorRoot, 32);
  if (!isCanonicalField(action.anchorRoot)) throw new Error('Anchor root is not canonical');
  for (const [index, nullifier] of action.nullifiers.entries()) {
    requireLength(`Nullifier ${index}`, nullifier, 32);
    if (!isCanonicalField(nullifier)) throw new Error(`Nullifier ${index} is not canonical`);
  }
  if (action.outputs.length !== 3) throw new Error('Action must contain exactly three outputs');
  for (const [index, output] of action.outputs.entries()) {
    requireLength(`Output ${index} commitment`, output.cm, 32);
    requireLength(`Output ${index} recipient envelope`, output.recipientEnvelope, 181);
    requireLength(`Output ${index} outgoing envelope`, output.outgoingEnvelope, 157);
    if (!isCanonicalField(output.cm)) throw new Error(`Output ${index} commitment is not canonical`);
    if (isZero(output.cm) || isZero(output.recipientEnvelope) || isZero(output.outgoingEnvelope)) {
      throw new Error(`Output ${index} package must be nonzero`);
    }
  }
  for (let left = 0; left < action.outputs.length; left += 1) {
    for (let right = left + 1; right < action.outputs.length; right += 1) {
      if (equalBytes(action.outputs[left].cm, action.outputs[right].cm)) {
        throw new Error('Output commitments must differ');
      }
    }
  }
  if (action.publicValue < 0n || action.publicValue > MAX_PUBLIC_VALUE) {
    throw new Error('Invalid public value');
  }

  const anchorIsZero = isZero(action.anchorRoot);
  if (action.nullifiers.some(isZero)) throw new Error('Nullifiers must be nonzero');
  if (equalBytes(action.nullifiers[0], action.nullifiers[1])) {
    throw new Error('Nullifiers must differ');
  }
  if (action.kind === ActionKind.Deposit) {
    if (!anchorIsZero) {
      throw new Error('Invalid deposit private slots');
    }
    if (
      action.publicValue === 0n ||
      !action.asset ||
      action.assetIndex === undefined ||
      !action.depositSource ||
      action.publicRecipient
    ) {
      throw new Error('Invalid deposit public boundary');
    }
    validateAddress(action.depositSource, 'Deposit source');
    return;
  }

  if (anchorIsZero) throw new Error('Invalid private spend slots');
  if (action.kind === ActionKind.PrivateTransfer) {
    if (
      action.publicValue !== 0n
      || action.asset
      || action.assetIndex !== undefined
      || action.depositSource
      || action.publicRecipient
    ) {
      throw new Error('Private transfer must not expose a boundary asset or address');
    }
    return;
  }
  if (
    action.publicValue === 0n
    || !action.asset
    || action.assetIndex === undefined
    || action.depositSource
    || !action.publicRecipient
  ) {
    throw new Error('Invalid withdrawal public boundary');
  }
  validateAddress(action.publicRecipient, 'Public recipient');
}

export function serializeCanonicalActionBytes(
  action: ActionModel,
  networkId: Uint8Array,
  realmId: Uint8Array,
  poolId: Uint8Array,
): Uint8Array {
  validateAction(action);
  for (const [name, bytes] of [
    ['Network ID', networkId],
    ['Realm ID', realmId],
    ['Pool ID', poolId],
  ] as const) {
    requireLength(name, bytes, 32);
  }
  const buf: number[] = [];
  encodeDomain(DOMAIN_ACTION, buf);
  encodeU16Be(action.protocolVersion, buf);
  for (const b of networkId) buf.push(b);
  for (const b of realmId) buf.push(b);
  for (const b of poolId) buf.push(b);

  buf.push(action.kind);
  encodeOptionalAddress(action.asset, buf);
  encodeU32Be(action.assetIndex ?? 0, buf);
  for (const b of action.actionNonce) buf.push(b);
  for (const b of action.anchorRoot) buf.push(b);
  for (const b of action.nullifiers[0]) buf.push(b);
  for (const b of action.nullifiers[1]) buf.push(b);

  // outputs[0] (370 bytes)
  for (const b of action.outputs[0].cm) buf.push(b);
  for (const b of action.outputs[0].recipientEnvelope) buf.push(b);
  for (const b of action.outputs[0].outgoingEnvelope) buf.push(b);

  for (let index = 1; index < 3; index += 1) {
    for (const b of action.outputs[index].cm) buf.push(b);
    for (const b of action.outputs[index].recipientEnvelope) buf.push(b);
    for (const b of action.outputs[index].outgoingEnvelope) buf.push(b);
  }

  encodeU64Be(action.publicValue, buf);
  encodeOptionalAddress(action.depositSource, buf);
  encodeOptionalAddress(action.publicRecipient, buf);

  return Uint8Array.from(buf);
}

export function computeAssetField(asset: { kind: number; payload: Uint8Array }): Uint8Array {
  validateAddress(asset, 'Asset');
  if (asset.kind !== 1) throw new Error('Asset must be a contract address');
  const bytes = new Uint8Array(33);
  bytes[0] = asset.kind;
  bytes.set(asset.payload, 1);
  return fieldId(DOMAIN_ASSET, bytes);
}

export function computeActionField(
  action: ActionModel,
  networkId: Uint8Array,
  realmId: Uint8Array,
  poolId: Uint8Array,
): Uint8Array {
  const bytes = serializeCanonicalActionBytes(action, networkId, realmId, poolId);
  return fieldId(DOMAIN_ACTION, bytes);
}

export async function computePublicSignals(
  action: ActionModel,
  contextField: Uint8Array,
  networkId: Uint8Array,
  realmId: Uint8Array,
  poolId: Uint8Array,
): Promise<Uint8Array[]> {
  validateAction(action);
  const assetField = action.asset ? computeAssetField(action.asset) : new Uint8Array(32);
  const actionField = computeActionField(action, networkId, realmId, poolId);

  const kindField = new Uint8Array(32);
  kindField[31] = action.kind;

  const valField = bigintTo32Bytes(action.publicValue);
  return [
    contextField,
    assetField,
    kindField,
    action.anchorRoot,
    valField,
    actionField,
    action.nullifiers[0],
    action.nullifiers[1],
    action.outputs[0].cm,
    action.outputs[1].cm,
    action.outputs[2].cm,
  ];
}
